File: task2/src/image_catch/scripts/image_catch_node.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

class ImageReceiver:
    def __init__(self):
        rospy.init_node("image_catch_node")
        rospy.loginfo("Node_init done")
        self.image_sub = rospy.Subscriber("usb_cam/image_raw",Image,self.image_callback,queue_size=10)
        self.image_pub = rospy.Publisher("image_process",Image,queue_size=10)
        self.image_count = 0
        self.max_count = 3  

        
    def image_callback(self, msg):
        bridge = CvBridge()
        cv_image = bridge.imgmsg_to_cv2(msg,"passthrough")
        cv_image_rgb = cv2.cvtColor(cv_image,cv2.COLOR_BGR2RGB)
        processed_msg = bridge.cv2_to_imgmsg(cv_image_rgb,"rgb8")
        self.image_pub.publish(processed_msg)
        self.image_count += 1
        rospy.loginfo("got one")
        if self.image_count >= self.max_count:
            rospy.logwarn("Received {} images. Shutting down...".format(self.max_count))
            rospy.signal_shutdown("Received enough images")
        else:
            rospy.sleep(5)

if __name__ == '__main__':
    node = ImageReceiver()
    rospy.spin()

File: task2/src/image_catch/src/image_catch_node.cpp
#include "ros/ros.h"
#include "sensor_msgs/Image.h"
#include "cv_bridge/cv_bridge.h"
#include "opencv2/opencv.hpp"

class ImageReceiver {
public:
    ImageReceiver() {
        ros::NodeHandle nh;
        ROS_INFO("Node_init done");
        image_sub = nh.subscribe("usb_cam/image_raw", 10, &ImageReceiver::imageCallback, this);
        image_pub = nh.advertise<sensor_msgs::Image>("image_process", 10);
        image_count = 0;
        max_count = 3;
    }

    void imageCallback(const sensor_msgs::ImageConstPtr& msg) {
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        } catch (cv_bridge::Exception& e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        cv::Mat cv_image_rgb;
        cv::cvtColor(cv_ptr->image, cv_image_rgb, cv::COLOR_BGR2RGB);

        sensor_msgs::ImagePtr processed_msg = cv_bridge::CvImage(std_msgs::Header(), "rgb8", cv_image_rgb).toImageMsg();
        image_pub.publish(processed_msg);

        image_count++;
        ROS_INFO("got one");
        if (image_count >= max_count) {
            ROS_WARN("Received %d images. Shutting down...", max_count);
            ros::shutdown();
        } else {
            ros::Duration(5).sleep();
        }
    }

private:
    ros::Subscriber image_sub;
    ros::Publisher image_pub;
    int image_count;
    int max_count;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "image_catch_node");
    ImageReceiver receiver;
    ros::spin();
    return 0;
}